Emit task returning thunks in crossgen2

jtschuster opened the pull request:
Emits Task-returning async method thunks in crossgen2 as MethodSignatures with the AsyncVariant flag set

Adds `GetPrimaryMethodDesc` to get the most natural "primary" method desc for MethodDesc's that point to the same Ecma/Instantiated method. This is used in the places where we cast to EcmaMethod to get relevant metadata information.

GCRefMap emission also needed to be updated to account for the continuation parameter. Runtime's `FakeGcScanRoots` (used to verify the ReadyToRun images GCRefMap in debug config) also needed to updated to handle the continuation parameter.

In R2R exception info clauses, a new flag is added to use when the handler catches System.Exception. It's not guaranteed that a token for System.Exception exists in the context of the EH info struct, and there is no equivalent of a module override for the structs.

In the MethodWithToken constructor, it's expected that callers pass the Token that points to the exact method that is being called on the correct instantiated type. It's again not guaranteed that we have valid tokens for the method being called in the context of an instantiated type (technically any call to `G<MyClass>.M()` will have a token for G<Class>, but when we compile the canonicalize method, we call the async `G<!0>.M()` in the IL, and we don't have a token for `G<!0>`). For this reason, we wrap the MethodIL in a ManifestModuleWrappedMethodIL, and when we construct a MethodWithToken, we resolve the OwningType with the MutableModule token, but swap out the Token for the typical method definition token and force emission of a full MethodSignature in the Fixup.

jtschuster committed:
Handle continuation parameter in R2R GCRefMaps

jtschuster committed:
Add helper extension methods

jtschuster committed:
Emit Task-returning thunks in crossgen2

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into MethodWithToken

jtschuster committed:
Undo additional changes

jtschuster committed:
Ensure type doesn't version with module group before finding external token

jtschuster commented:
/azp run runtime-coreclr crossgen2 outerloop

davidwrighton commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:
```diff
@@ -558,6 +558,9 @@ public static bool ShouldSkipCompilation(InstructionSetSupport instructionSetSup
 
         public static bool ShouldCodeNotBeCompiledIntoFinalImage(InstructionSetSupport instructionSetSupport, MethodDesc method)
         {
+            if (method.IsAsyncVariant())
```
This should at least respect the BypassReadyToRunAttribute.

jtschuster replied:
Updated to use GetPrimaryMethodDesc to get the EcmaMethod.

jtschuster committed:
Punt on generic async thunks, skip all async-variant methods

jtschuster committed:
Get PrimaryMethodDesc for ShouldCodeNotBeCompiled

jtschuster commented:
/azp run runtime-coreclr crossgen2 outerloop

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into EmitTaskReturningThunks

jtschuster committed:
Skip non-task-returning async methods instead of throwing.

jtschuster committed:
Use an unconditionally true filter clause instead of catching object

The generated IL thunk does not have valid tokens, so the tokens emitted in the EH handler info are invalid. It's not an invariant that there is always a reference to System.Exception or System.Object in the assembly that  we are generating thunks for, so we can't always get a valid Token for the exception type token field. Instead, we can create a filter that does a type check against System.Exception. The reference to System.Exception will be filled in with a Fixup with a module override if we don't have a reference to System.Exception in the assembly.

jtschuster commented:
/azp run runtime-coreclr crossgen2 outerloop

jtschuster committed:
Use class token EHInfo in NAOT, filter in R2R

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into EmitTaskReturningThunks

jkotas commented on src/coreclr/tools/Common/TypeSystem/IL/Stubs/AsyncThunks.cs:
```diff
@@ -142,11 +142,20 @@ public static MethodIL EmitTaskReturningThunk(MethodDesc taskReturningMethod, Me
+                    codestream.EndTry(tryFilterRegion);
                 }
-                // Catch
+                // Filter - check if the caught object is a System.Exception
```
The token in EH clause from R2R image is resolved here:

https://github.com/dotnet/runtime/blob/3a11eaf96847b9804c6d4706a0cbd14903eb7343/src/coreclr/vm/codeman.cpp#L6494-L6500

We just need to ensure that the encoding produced by crossgen2 can be resolved in this method.

> Or we could make a class token of 0 represent "catch anything".

Yes, it would be an option. There is an interesting detail in how to handle exceptions that do not inherit from System.Exception. These exceptions may or may not get wrapped as `RuntimeWrappedException` depending on per-assembly settings. What is the behavior we want for these stubs? Do we want them to be wrapped unconditionally to compat with async v1?

I see that you have added code in the filter to specifically handle this case. Do we have any test for it?

davidwrighton replied:
The goal here is to match async v1 behavior. This implies that we need to test/validate the behavior of traditional async with regards to the RuntimeWrappedException behavior and match up to whatever it is (for cases where the async method is defined in an assemblies with both of the relevant options. Sounds like a test hole we need to fill.

jtschuster replied:
I don't think I fully understand all the implications of `WrapNonExceptionThrows` on async v1 vs async v2. Do we want a passing test suite as a prereq for merging this PR? And do you have a preference for adding a `CorExceptionFlag` or treating `ClassToken == 0` as `System.Exception`, or finding another solution? I lean towards using a new CorExceptionFlag for R2R.

jtschuster committed:
Emit null token to represent catching a System.Exception

With these changes, we no longer need separate IL for nativeaot and readytorun

jtschuster committed:
Undo extra ClassToken == 0 check

jtschuster committed:
Add the emission logic in crossgen

davidwrighton committed:
Rework async thunks generation to cross module generics infra (#4)

* Rework async thunks generation to cross module generics infra
- Add ability for code generating an ILStub to mark that the tokens might be generated
- Funnel through various details to ensure that we trigger creation of the ManifestModuleWrappedMethodIL as needed
- Adjust method fixup signature generation to allow for VersionsWithMethodBody code which is in the MutableModule
- Add a few todos about work we might do to allow cross module compilation of runtime async code.
  - Disable this for now though, since its much to complex to test without a working end-to-end scenario (See code in CrossModuleInlineableUncached

* Revert logic which attempts to treat System.Exception in EH clause specially

jtschuster committed:
Update src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

jtschuster committed:
Encode/decode ClassToken of 0 as System.Exception for async thunks

jtschuster committed:
Use COR_ILEXCEPTION_CLAUSE_R2R_SYSTEM_EXCEPTION flag instead of ClassToken == 0

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into EmitTaskReturningThunks

jtschuster committed:
Remove unneeded code

jtschuster committed:
Use MutableModule token for resolving owning type, def token for signature

jtschuster committed:
Remove aot.slnx

jtschuster commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs:
```diff
@@ -700,51 +700,6 @@ protected override void ComputeDependencyNodeDependencies(List<DependencyNodeCor
                         }
-                        if (method.IsAsyncThunk())
-                        {
-                            // The synthetic async thunks require references to methods/types that may not have existing methodRef entries in the version bubble.
```
Async thunks are now wrapped in ManifestModuleWrappedMethodIL, which will add the references needed to the MutableModule when they are initialized during the single threaded portion of the compilation. It turns out this code was duplicating work that @davidwrighton already implemented and isn't needed anymore.

jtschuster committed:
Address PR feedback for R2R_SYSTEM_EXCEPTION EH clause encoding

- Clear ClassTokenOrOffset when setting R2R_SYSTEM_EXCEPTION flag
- Add COR_ILEXCEPTION_CLAUSE_R2R_SYSTEM_EXCEPTION to R2R dumper

jtschuster committed:
Merge branch 'main' into EmitTaskReturningThunks

jtschuster committed:
Merge branch 'main' into EmitTaskReturningThunks
